### ISSUE TITLE
Setting useBestModel on NerDLApproach with a test dataset causes data leakage in training

### ISSUE THREAD

**Issue by ethnhll** [JohnSnowLabs/spark-nlp#14033]
### Is there an existing issue for this?

- [X] I have searched the existing issues and did not find a match.

### Who can help?

@maziyarpanahi @vkocaman 

### What are you working on?

Training dl NER  models.

### Current Behavior

Currently, when setting `useBestModel` along with `setTestDataset`, data leakage from the test set into training occurs.

when `useBestModel` is set, after each epoch the F1 score is calculated based on the test dataset and a candidate model (lastCheckPoints) is set based on this code starting at [line 314 of TensorflowNer.scala](https://github.com/JohnSnowLabs/spark-nlp/blob/1fa94e90325366d100c0ab26a28600f66b2e52ab/src/main/scala/com/johnsnowlabs/ml/tensorflow/TensorflowNer.scala#L314)
```scala
      
    if (test.nonEmpty) {
        println("Quality on test dataset: ")
        outputLog("Quality on test dataset: ", uuid, enableOutputLogs, outputLogsPath)
        val (newTestMicroF1, newTestMacroF1) = measure(
          test,
          extended = evaluationLogExtended,
          enableOutputLogs = enableOutputLogs,
          outputLogsPath = outputLogsPath,
          batchSize = batchSize,
          uuid = uuid)
        if (useBestModel && bestModelMetric == ModelMetrics.testMicroF1) {
          if (newTestMicroF1 >= lastTestMircoF1) {
            lastCheckPoints = saveBestModel()
            lastTestMircoF1 = newTestMicroF1
          }
        }
      }
```
This determines which epoch produced the “best model”, ie the test dataset made a deciding influence on a hyper parameter of the model in training, rather than the 

It also looks like even if one were to use `setBestModelMetric` to set the metric to something like `valMicroF1`, this setting is ignored/overwritten both in `NerDLApproach.scala` and in `TensorflowNer.scala`.

Based on `NerDLApproach.scala` on [line 437](https://github.com/JohnSnowLabs/spark-nlp/blob/1fa94e90325366d100c0ab26a28600f66b2e52ab/src/main/scala/com/johnsnowlabs/nlp/annotators/ner/dl/NerDLApproach.scala#L437), `bestModelMetric` can only ever be set to `ModelMetrics.macroF1` or `ModelMetrics.microF1`.

And in `TensorflowNer.scala` starting on [line 198](https://github.com/JohnSnowLabs/spark-nlp/blob/1fa94e90325366d100c0ab26a28600f66b2e52ab/src/main/scala/com/johnsnowlabs/ml/tensorflow/TensorflowNer.scala#L198), `bestModelMetric` is overwritten with `ModelMetrics.testMicroF1` when a non-empty test dataset is provided.

### Expected Behavior

I would expect that the validation split from training data would be used to determine which epoch produced the best model after training, and only after training is completed would the model get evaluated against the test set, ie the test set should have no influence on which epoch produced the best model.

### Steps To Reproduce

```python

graph_builder = (
    TFGraphBuilder()
    .setInputCols(input_columns)
    .setLabelColumn("ner_label")
    .setGraphFolder("./ner_graphs")
    .setGraphFile("auto")
    .setHiddenUnitsNumber(24)
    .setModelName("ner_dl")
)

ner_approach = (
    NerDLApproach()
    .setInputCols(input_columns)
    .setLabelColumn('ner_label')
    .setOutputCol("named_entity")
    .setGraphFolder("./ner_graphs")
    .setValidationSplit(.20)
    .setBatchSize(64)
    .setLr(0.003)
    .setMaxEpochs(50)
    .setRandomSeed(42)
    .setTestDataset(test_dataset)
    .setUseBestModel(True)
)

pipeline = Pipeline().setStages([graph_builder, ner_approach])
training_set = spark.read.format('parquet').load(training_set_path)

trained_model = pipeline.fit(training_set)
```

### Spark NLP version and Apache Spark

sparknlp.version 5.1.2
spark.version 3.4.1

### Type of Spark Application

Python Application

### Java Version

Java: Zulu 8.70.0.23-CA-linux64

### Additional Information

run on Databricks DBR 13.3 LTS

**Comment by maziyarpanahi** [comment]
Thanks @ethnhll 

I will refactor this part to first:
- add the missing tracking for validation dataset - in this case we will track all 3 test, validation split, and validation dataset if they are all exist
- not override `setBestModelMetric` and allow user to set the desired metric to override the bestModel. (it seems currently, this is based on some order and ignores the actual test/val/split and only cares about Micro or Macro)

Will include the fix in the next release

**Comment by ethnhll** [comment]
I don't think test metric should be used to select the best model, because that is what is causing the leakage.

**Comment by maziyarpanahi** [comment]
I don't think it should be up to us to decide that. We allow users to select among these 3 available metrics to save the best model:

- validation dataset (if exists)
- test dataset (if exists)
- validation split (%, if exists)

People use different datasets for validation or test, we cannot make that decision for them. This bug aside, once we refactor this part and give users the true power of setting the metrics, it should be up to them which one they like. If `test` is not a good metrics for choosing the best model, they can choose `validation`. Either way all the metrics will be printed out and logged. And only the chosen one will be used to save the best model. 

(if we did not have `bestModelMetric` parameter, then it's true, we would go with validation dataset if it exists. But we have it, and once it's fixed users can use it to decide what to be used)

**Comment by ethnhll** [comment]
I suppose that's fair, though I would argue the library shouldn't make it easy for a user to shoot themselves in the foot by accident, so a refactor should make the defaults take the correct path through those choices (ie valMetricf1 should be default even if a test dataset is used, because that is what would be technically correct).
